rtloader/three/three.cpp

```diff
@@ -60,45 +63,32 @@ Three::~Three()
 
 void Three::initPythonHome(const char *pythonHome)
 {
-    // Py_SetPythonHome stores a pointer to the string we pass to it, so we must keep it in memory
-    wchar_t *oldPythonHome = _pythonHome;
-    if (pythonHome == NULL || strlen(pythonHome) == 0) {
-        _pythonHome = Py_DecodeLocale(_defaultPythonHome, NULL);
-    } else {
-        _pythonHome = Py_DecodeLocale(pythonHome, NULL);
+    const auto home = pythonHome != nullptr && strlen(pythonHome) > 0 ? pythonHome : _defaultPythonHome;
+    const auto status = PyConfig_SetBytesString(&_config, &_config.home, home);
+    if (PyStatus_Exception(status)) {
+        setError("Failed to set python home");
     }
-
-    Py_SetPythonHome(_pythonHome);
-    PyMem_RawFree((void *)oldPythonHome);
 }
 
 void Three::initPythonExe(const char *python_exe)
 {
-    // Py_SetProgramName stores a pointer to the string we pass to it, so we must keep it in memory
-    wchar_t *oldPythonExe = _pythonExe;
-    _pythonExe = Py_DecodeLocale(python_exe, NULL);
-
-    Py_SetProgramName(_pythonExe);
-
-    // HACK: This extra internal API invocation is due to the workaround for an upstream bug on
-    // Windows (https://bugs.python.org/issue34725) where just using `Py_SetProgramName` is
-    // ineffective. The workaround API call will be removed at some point in the future (Python
-    // 3.12+) so we should convert this initialization to the new`PyConfig API`
-    // (https://docs.python.org/3.11/c-api/init_config.html#c.PyConfig) before then.
-    _Py_SetProgramFullPath(_pythonExe);
-
-    PyMem_RawFree((void *)oldPythonExe);
+    const auto status = PyConfig_SetBytesString(&_config, &_config.executable, python_exe);
+    if (PyStatus_Exception(status)) {
+        setError("Failed to set python executable");
+    }
 }
 
 bool Three::init()
 {
     // we want the checks to be runned with the standard encoding utf-8
-    // setting this var to 1 forces the UTF8 mode for CPython >= 3.7
-    // See:
-    //	- PEP UTF8 mode https://www.python.org/dev/peps/pep-0540/
-    //	- about this var https://github.com/python/cpython/pull/12589
-    // This has to be set before the Py_Initialize() call.
-    Py_UTF8Mode = 1;
+    PyPreConfig preConfig;
+    PyPreConfig_InitPythonConfig(&preConfig);
```

Comment:
We call this, but pre-init would be done earlier implicitly by a call to `PyConfig_SetBytesString`. Is this still effective?

We may install a custom allocator in `initPymemStats`, which, according to the [documentation](https://docs.python.org/3/c-api/init_config.html#c.Py_PreInitialize#:~:text=PyMem_SetAllocator) should be done after pre-init. In general, we call a bunch of stuff before `init` (like registering agent's api modules), are those safe to do before pre-init?

Reply:
https://docs.python.org/3/c-api/init_config.html#c.Py_PreInitializeFromArgs mentions `PyMem_SetAllocator() can be called after [Py_PreInitialize()](https://docs.python.org/3/c-api/init_config.html#c.Py_PreInitialize) and before [Py_InitializeFromConfig()](https://docs.python.org/3/c-api/init.html#c.Py_InitializeFromConfig) to install a custom memory allocator. It can be called before [Py_PreInitialize()](https://docs.python.org/3/c-api/init_config.html#c.Py_PreInitialize) if [PyPreConfig.allocator](https://docs.python.org/3/c-api/init_config.html#c.PyPreConfig.allocator) is set to PYMEM_ALLOCATOR_NOT_SET.` so I think as far as allocators are concerned, we're good.

I'm not sure I'm knowledgeable enough about rtloader to comment in a more general manner.

AFAICS any later preinit would be ignored, so we'd need to move this to an ealier point, most likely as part of the constructor